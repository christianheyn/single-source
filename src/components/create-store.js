import { deepcopy } from './deepcopy';
import { extendStateObj } from './extend-state-obj';
import dotProp from 'dot-prop';

export const createStore = (initialState = {}) => {
    const store = { state: deepcopy(initialState) };
    const subscribeCollection = [];

    return {
        dispatch: (actionObj) => {
            store.state = extendStateObj(store.state, actionObj);
        },
        subscribe: (path = '', callback) => subscribeCollection.push({ path, callback }),
        getState: (path = '') => {
            if (typeof path !== 'string' || path === '') {
                return deepcopy(store.state);
            }

            const partOfObject = dotProp.get(store.state, path);
            const typeofPart = typeof partOfObject;

            return (typeofPart === 'object')
                ? deepcopy(partOfObject)
                : (typeofPart !== 'undefined')
                ? partOfObject
                : undefined;
        },
    };
};
