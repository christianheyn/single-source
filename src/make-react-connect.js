const isReact163 = versionString => !!versionString.match(/^16\.3/);
const createObjectFromStore = (store, mappingObj) => {
    const objKeys = Object.keys(mappingObj);
    const resultObj = {};
    objKeys.forEach((propKey) => {
        resultObj[propKey] = store.getState(mappingObj[propKey]);
    });

    return resultObj;
};

export const makeReactConnect = (React, store, mappingObj) => (Component) => {
    const Provider = (props) => {
        const $ = {
            ...React.Component.prototype,
            props,
            state: {},
        };

        const privates = {
            state: createObjectFromStore(store, mappingObj),
        };

        const objKeys = Object.keys(mappingObj);
        objKeys.forEach((propKey) => {
            const path = mappingObj[propKey];
            store.subscribe(path, (newData) => {
                privates.state[propKey] = newData;
                $.forceUpdate();
            });
        });

        // TODO:
        /*
        if (!isReact163(React.version)) {
            $.componentWillReceiveProps = (nextProps) => {
                // props to state
            }
        }
        */

        $.render = () => {
            const { state } = $;
            const resultProps = {
                ...state,
                ...privates.state,
            };

            return <Component {...resultProps} />;
        };

        return $;
    };

    if (isReact163(React.version)) {
        Provider.getDerivedStateFromProps = (nextProps) => {
            return nextProps;
        };
    }

    return Provider;
};
