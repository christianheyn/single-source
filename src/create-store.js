import { deepCopy, deepEqual } from './components/deep';
import { extendStateObj } from './components/extend-state-obj';
import { dotPropGet } from './components/dot-prop';
import { filterSubscriptions } from './components/filter-subscriptions';
import { unsubscribe } from './components/unsubscribe';

export const createStore = (initialState = {}) => {
    const privates = {
        state: deepCopy(initialState),
        subscriptionCollection: [],
    };

    const getState = (path = '') => {
        if (typeof path !== 'string' || path === '') {
            return deepCopy(privates.state);
        }

        const partOfObject = dotPropGet(privates.state, path);
        const typeofPart = typeof partOfObject;

        return (typeofPart === 'object')
            ? deepCopy(partOfObject)
            : (typeofPart !== 'undefined')
            ? partOfObject
            : undefined;
    };

    const dispatch = (actionObj) => {
        const lastState = deepCopy(privates.state);
        const newState = extendStateObj(privates.state, actionObj);

        if (deepEqual(lastState, newState)) {
            return undefined;
        }

        privates.state = newState;

        const filteredActionObjs = filterSubscriptions(
            privates.subscriptionCollection,
            lastState,
            newState,
        );

        if (filteredActionObjs.length === 0) {
            return undefined;
        }

        filteredActionObjs.forEach((action) => {
            const changedData = dotPropGet(privates.state, action.path);
            action.callback(changedData);
        });

        return undefined;
    };

    const subscribe = (path = '', callback) => {
        const now = Date.now();
        privates.subscriptionCollection.push({ path, callback, id: now });
        return { id: now };
    };

    const unsubscribeMethod = (subscriptionObj) => {
        privates.subscriptionCollection = unsubscribe(
            privates.subscriptionCollection,
            subscriptionObj,
        );

        return undefined;
    };

    return {
        dispatch,
        subscribe,
        unsubscribe: unsubscribeMethod,
        getState,
    };
};
