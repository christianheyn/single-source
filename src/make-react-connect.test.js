import React from 'react';
import { create } from 'react-test-renderer';
import { makeReactConnect } from './make-react-connect';
import { createStore } from './create-store';

const TestComponent = (props) => {
    const $ = {
        ...React.Component.prototype,
        props,
    };

    $.render = () => {
        const { text, headline } = $.props;

        return (
            <div className={'test-component'}>
                <h1>{ headline }</h1>
                { text }
            </div>
        );
    };

    return $;
};

describe('makeReactConnect(React, )', () => {
    it('is a function', () => {
        const actual = typeof makeReactConnect;
        const expected = 'function';
        expect(actual).toEqual(expected);
    });

    it('returns function', () => {
        const actual = typeof makeReactConnect();
        const expected = 'function';
        expect(actual).toEqual(expected);
    });

    it('--- no connection ---', () => {
        const props = {
            headline: 'Hello World',
            text: 'no text yet',
        };
        const actual = create(<TestComponent {...props} />);
        expect(actual).toMatchSnapshot();
    });

    it('returns connected component', () => {
        const testStore = createStore({ testData: { text: 'HelloWorld!' } });
        const ConnectedComponent =
            makeReactConnect(React, testStore, {
                text: 'testData.text',
            })(TestComponent);
        const props = {
            headline: 'Hello World',
            text: 'no text yet',
        };
        const actual = create(<ConnectedComponent {...props} />);
        expect(actual).toMatchSnapshot();
    });

    it('rerenders connected component when dispatch', () => {
        const testStore = createStore({ testData: { text: 'HelloWorld!' } });
        const ConnectedComponent =
            makeReactConnect(React, testStore, {
                text: 'testData.text',
            })(TestComponent);
        const props = {
            headline: 'Hello World',
            text: 'no text yet',
        };
        testStore.dispatch({
            path: 'testData.text',
            payload: 'refreshed text!',
        });

        const actual = create(<ConnectedComponent {...props} />);
        expect(actual).toMatchSnapshot();
    });
});
