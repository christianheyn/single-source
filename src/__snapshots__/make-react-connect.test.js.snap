// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeReactConnect(React, ) --- no connection --- 1`] = `
<div
  className="test-component"
>
  <h1>
    Hello World
  </h1>
  no text yet
</div>
`;

exports[`makeReactConnect(React, ) rerenders connected component when dispatch 1`] = `
<div
  className="test-component"
>
  <h1>
    Hello World
  </h1>
  refreshed text!
</div>
`;

exports[`makeReactConnect(React, ) returns connected component 1`] = `
<div
  className="test-component"
>
  <h1>
    Hello World
  </h1>
  HelloWorld!
</div>
`;
